{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b17e7bd-1aae-499d-9bcc-b74dc07c4719",
   "metadata": {},
   "outputs": [],
   "source": [
    "import dask.bag as db\n",
    "from glob import glob\n",
    "import pandas as pd\n",
    "import re\n",
    "import numpy as np\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import plotly.express as px\n",
    "import plotly.graph_objs as go"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "559bf5b8-0d92-4db6-aac4-8c8eb80689de",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "This code snippet and Data is courtesy of Thiago Sanches from Palmer Lab\n",
    "'''\n",
    "\n",
    "# allgwas = glob(f'''/tscc/projects/ps-palmer/gwas/projects/*/results/gwas/*_chrgwas*.mlma''') \n",
    "#### here in the second * you can specify a chromosome or a list of chromosomes this one will load everything\n",
    "\n",
    "target_1 = pd.read_csv('target_1.txt', header = None, names=['SNP'])\n",
    "target_1 = list(target_1.SNP)\n",
    "\n",
    "allgwas = glob(f'''/tscc/projects/ps-palmer/gwas/projects/*/results/gwas/*_chrgwas1.mlma''') #### example for chr 5 \n",
    "seq = db.from_sequence(allgwas[:])\n",
    "gwas_reader = lambda x: pd.read_csv(x, dtype={'Chr':int, 'SNP':str, 'bp':int}, sep = '\\t')\\\n",
    "                              .assign(trait = x.split('regressedlr_')[-1].split('_chrgwas')[0],\n",
    "                                      project = re.findall('gwas/projects/(.*)/results', x)[0],\n",
    "                                     )\\\n",
    "                              .set_index('SNP')\n",
    "# .reindex(target_1)\n",
    "#### you can subset in this reader function to speed up the code too, the index is the snp names, so if you have a list you can use that\n",
    " ### if you want to subset at this point\n",
    "allgwas_res = pd.concat(seq.map(gwas_reader).compute())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed6938e3-a7bf-4ab1-8385-993d503ed9b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "allgwas_res['-log10 P-Value'] = -np.log10(allgwas_res['p'])\n",
    "\n",
    "# allgwas_res_sorted = allgwas_res.sort_values(by='-log10 P-Value', ascending=False)\n",
    "\n",
    "# Drop duplicates based on the 'trait' column, keeping the first occurrence (largest p-value)\n",
    "# allgwas_res_no_duplicates = allgwas_res_sorted.drop_duplicates(subset='trait', keep='first')\n",
    "\n",
    "# Get the top 20 traits with the largest p-values (if you have more than 20)\n",
    "# top_20_traits = allgwas_res_no_duplicates.nlargest(20, '-log10 P-Value')\n",
    "\n",
    "# li = top_20_traits.trait\n",
    "\n",
    "# phenotype = allgwas_res[allgwas_res.trait.isin(li)]\n",
    "\n",
    "# phenotype.bp = phenotype.bp.astype(int)\n",
    "# phenotype.Chr = phenotype.Chr.astype(int)\n",
    "# append_position = phenotype.groupby('Chr').bp.agg('max').sort_index().cumsum().shift(1,fill_value=0)\n",
    "# phenotype['Chromosome'] = phenotype.apply(lambda row: int(row.bp) + append_position[row.Chr], axis = 1)\n",
    "# phenotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d3bd3e4-fb8a-4673-b372-08f58b82d449",
   "metadata": {},
   "outputs": [],
   "source": [
    "allgwas_res.bp = allgwas_res.bp.astype(int)\n",
    "allgwas_res.Chr = allgwas_res.Chr.astype(int)\n",
    "append_position = allgwas_res.groupby('Chr').bp.agg('max').sort_index().cumsum().shift(1,fill_value=0)\n",
    "allgwas_res['Chromosome'] = allgwas_res.apply(lambda row: int(row.bp) + append_position[row.Chr], axis = 1)\n",
    "allgwas_res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f046c6b-9f09-4749-bc38-969311ffcaee",
   "metadata": {},
   "outputs": [],
   "source": [
    "#position the x-axis by substituting the distance with positions in bp\n",
    "\n",
    "li = top_20_traits.trait\n",
    "\n",
    "phenotype = allgwas_res[allgwas_res.trait.isin(li)]\n",
    "\n",
    "phenotype.bp = phenotype.bp.astype(int)\n",
    "phenotype.Chr = phenotype.Chr.astype(int)\n",
    "append_position = phenotype.groupby('Chr').bp.agg('max').sort_index().cumsum().shift(1,fill_value=0)\n",
    "phenotype['Chromosome'] = phenotype.apply(lambda row: int(row.bp) + append_position[row.Chr], axis = 1)\n",
    "\n",
    "phenotype = phenotype.reset_index().rename(columns={'index': 'SNP'})\n",
    "\n",
    "highlighted_snp = '1:78033225'\n",
    "\n",
    "phenotype['highlight'] = phenotype.SNP.apply(lambda x: 'Highlighted' if x == highlighted_snp else 'Normal')\n",
    "\n",
    "# Create a scatter plot with phenotypes on x and p-values on y\n",
    "fig = px.scatter(\n",
    "    phenotype, x='bp', y='-log10 P-Value',\n",
    "    hover_name=phenotype.index,  # Show SNP names on hover\n",
    "    title='Target SNP: 1:78033225 (3MB surround)',\n",
    "    color = 'trait',\n",
    "    size=phenotype['SNP'].apply(lambda x: 20 if x == highlighted_snp else 5),\n",
    "    labels={'SNP': 'SNP', '-log10 P-Value': '-log10 P-Value'},  # Color points by phenotype\n",
    ")\n",
    "\n",
    "fig.update_traces(marker=dict(line=dict(width=0, color='DarkSlateGrey')))\n",
    "\n",
    "fig.update_traces(\n",
    "    selector=dict(mode='markers'),  # Ensure we are customizing marker properties\n",
    "    marker=dict(\n",
    "        line=dict(\n",
    "            width=[4 if highlight == 'Highlighted' else 0 for highlight in phenotype['highlight']]\n",
    "        )\n",
    "    )\n",
    ")\n",
    "\n",
    "fig.update_layout(\n",
    "    width=1000,  # Set the width of the plot\n",
    "    height=800   # Set the height of the plot\n",
    ")\n",
    "\n",
    "# Show the interactive plot\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33a9a2cf-3ec9-4705-bb06-0dd832f5d7b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Manhattan plot of Chromosome 1 for Trait: kidney_left_g\n",
    "\n",
    "phenotype = allgwas_res[allgwas_res.trait == 'liver_left_weight']\n",
    "\n",
    "phenotype = phenotype.reset_index().rename(columns={'index': 'SNP'})\n",
    "\n",
    "highlighted_snp = '1:78033225'\n",
    "\n",
    "highest_snp = phenotype.loc[phenotype['-log10 P-Value'].idxmax(), 'SNP']\n",
    "\n",
    "phenotype['highlight'] = phenotype.SNP.apply(lambda x: 'Highlighted' if x == highlighted_snp else 'Normal')\n",
    "\n",
    "# Create a scatter plot with phenotypes on x and p-values on y\n",
    "fig = px.scatter(\n",
    "    phenotype, x='bp', y='-log10 P-Value',\n",
    "    hover_name=phenotype.index,  # Show SNP names on hover\n",
    "    title='Chrom 1, Target SNP: 1:78033225, Trait: liver_left_weight',\n",
    "    size=phenotype['SNP'].apply(lambda x: 10 if x == highlighted_snp else 2),  # Color points by phenotype\n",
    ")\n",
    "\n",
    "fig.update_traces(marker=dict(line=dict(width=0, color='DarkSlateGrey')))\n",
    "\n",
    "fig.update_traces(\n",
    "    selector=dict(mode='markers'),  # Ensure we are customizing marker properties\n",
    "    marker=dict(\n",
    "        line=dict(\n",
    "            width=[4 if highlight == 'Highlighted' else 0 for highlight in phenotype['highlight']]\n",
    "        )\n",
    "    )\n",
    ")\n",
    "\n",
    "target_idx = phenotype[phenotype['SNP'] == highlighted_snp].index[0]  # First occurrence of the target SNP\n",
    "fig.add_annotation(\n",
    "    x=phenotype['bp'][target_idx],\n",
    "    y=phenotype['-log10 P-Value'][target_idx],\n",
    "    text=f\"Target: {highlighted_snp}\",\n",
    "    showarrow=True,\n",
    "    arrowhead=2,\n",
    "    ax=0,\n",
    "    ay=-40,\n",
    "    bgcolor=\"yellow\"\n",
    ")\n",
    "\n",
    "# Annotate the highest SNP\n",
    "highest_idx = phenotype['-log10 P-Value'].idxmax()  # Index of the highest SNP\n",
    "fig.add_annotation(\n",
    "    x=phenotype['bp'][highest_idx],\n",
    "    y=phenotype['-log10 P-Value'][highest_idx],\n",
    "    text=f\"Highest: {highest_snp}\",\n",
    "    showarrow=True,\n",
    "    arrowhead=2,\n",
    "    ax=0,\n",
    "    ay=-40,\n",
    "    bgcolor=\"lightgreen\"\n",
    ")\n",
    "\n",
    "fig.update_layout(\n",
    "    width=1500,  # Set the width of the plot\n",
    "    height=800,   # Set the height of the plot\n",
    "    showlegend=False\n",
    ")\n",
    "\n",
    "# Show the interactive plot\n",
    "fig.write_image(\"fig5.png\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
